File: app/src/main/java/mega/privacy/android/app/main/managerSections/OutgoingSharesFragment.java
package mega.privacy.android.app.main.managerSections;

import static mega.privacy.android.app.main.ManagerActivity.OUTGOING_TAB;
import static mega.privacy.android.app.utils.Constants.OUTGOING_SHARES_ADAPTER;
import static mega.privacy.android.app.utils.Constants.VIEWER_FROM_OUTGOING_SHARES;
import static mega.privacy.android.app.utils.MegaNodeUtil.areAllFileNodesAndNotTakenDown;
import static mega.privacy.android.app.utils.SortUtil.sortByNameAscending;
import static mega.privacy.android.app.utils.SortUtil.sortByNameDescending;
import static mega.privacy.android.app.utils.Util.isScreenInPortrait;
import static nz.mega.sdk.MegaApiJava.INVALID_HANDLE;
import static nz.mega.sdk.MegaError.API_OK;
import static nz.mega.sdk.MegaShare.ACCESS_FULL;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.view.ActionMode;

import java.util.ArrayList;
import java.util.List;

import mega.privacy.android.app.R;
import mega.privacy.android.app.components.NewGridRecyclerView;
import mega.privacy.android.app.fragments.MegaNodeBaseFragment;
import mega.privacy.android.app.main.adapters.MegaNodeAdapter;
import mega.privacy.android.app.utils.CloudStorageOptionControlUtil;
import mega.privacy.android.app.utils.MegaNodeUtil;
import nz.mega.sdk.MegaApiJava;
import nz.mega.sdk.MegaError;
import nz.mega.sdk.MegaNode;
import nz.mega.sdk.MegaShare;
import timber.log.Timber;

public class OutgoingSharesFragment extends MegaNodeBaseFragment {

    @Override
    public void activateActionMode() {
        if (!adapter.isMultipleSelect()) {
            super.activateActionMode();

            if (getActivity() != null) {
                actionMode = ((AppCompatActivity) getActivity())
                        .startSupportActionMode(new ActionBarCallBack(OUTGOING_TAB));
            }
        }
    }

    @Override
    protected int viewerFrom() {
        return VIEWER_FROM_OUTGOING_SHARES;
    }

    private class ActionBarCallBack extends BaseActionBarCallBack {

        public ActionBarCallBack(int currentTab) {
            super(currentTab);
        }

        @Override
        public boolean onPrepareActionMode(ActionMode mode, Menu menu) {
            super.onPrepareActionMode(mode, menu);

            CloudStorageOptionControlUtil.Control control =
                    new CloudStorageOptionControlUtil.Control();

            if (selected.size() == 1 && !selected.get(0).isTakenDown()) {
                if (megaApi.checkAccessErrorExtended(selected.get(0), MegaShare.ACCESS_OWNER).getErrorCode()
                        == MegaError.API_OK) {
                    if (selected.get(0).isExported()) {
                        control.manageLink().setVisible(true)
                                .setShowAsAction(MenuItem.SHOW_AS_ACTION_ALWAYS);

                        control.removeLink().setVisible(true);
                    } else {
                        control.getLink().setVisible(true)
                                .setShowAsAction(MenuItem.SHOW_AS_ACTION_ALWAYS);
                    }
                }

                if (selected.get(0).isFolder()) {
                    control.shareFolder().setVisible(true)
                            .setShowAsAction(MenuItem.SHOW_AS_ACTION_ALWAYS);
                }

                if (selected.get(0).isOutShare()) {
                    control.removeShare().setVisible(true);
                }
            }

            boolean areAllNotTakenDown = MegaNodeUtil.areAllNotTakenDown(selected);

            if (areAllNotTakenDown) {
                if (managerActivity.getParentHandleOutgoing() == INVALID_HANDLE) {
                    control.removeShare().setVisible(true)
                            .setShowAsAction(MenuItem.SHOW_AS_ACTION_ALWAYS);
                }

                control.shareOut().setVisible(true)
                        .setShowAsAction(MenuItem.SHOW_AS_ACTION_ALWAYS);

                if (managerActivity.getDeepBrowserTreeOutgoing() > 0) {
                    if (areAllFileNodesAndNotTakenDown(selected)) {
                        control.sendToChat().setVisible(true)
                                .setShowAsAction(MenuItem.SHOW_AS_ACTION_ALWAYS);
                    }
                }

                control.copy().setVisible(true);
                if (control.alwaysActionCount() < CloudStorageOptionControlUtil.MAX_ACTION_COUNT) {
                    control.copy().setShowAsAction(MenuItem.SHOW_AS_ACTION_ALWAYS);
                } else {
                    control.copy().setShowAsAction(MenuItem.SHOW_AS_ACTION_NEVER);
                }
            } else {
                control.saveToDevice().setVisible(false);
            }

            if (selected.size() == 1
                    && megaApi.checkAccessErrorExtended(selected.get(0), ACCESS_FULL).getErrorCode() == API_OK) {
                control.rename().setVisible(true);
                if (control.alwaysActionCount() < CloudStorageOptionControlUtil.MAX_ACTION_COUNT) {
                    control.rename().setShowAsAction(MenuItem.SHOW_AS_ACTION_ALWAYS);
                } else {
                    control.rename().setShowAsAction(MenuItem.SHOW_AS_ACTION_NEVER);
                }
            }

            control.selectAll().setVisible(notAllNodesSelected());
            control.trash().setVisible(MegaNodeUtil.canMoveToRubbish(selected));

            CloudStorageOptionControlUtil.applyControl(menu, control);

            return true;
        }
    }

    public static OutgoingSharesFragment newInstance() {
        return new OutgoingSharesFragment();
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        super.onCreateView(inflater, container, savedInstanceState);

        Timber.d("onCreateView");

        if (megaApi.getRootNode() == null) {
            return null;
        }

        managerActivity.showFabButton();

        View v;

        if (managerActivity.isList) {
            v = getListView(inflater, container);

            if (adapter == null) {
                adapter = new MegaNodeAdapter(context, this, nodes,
                        managerActivity.getParentHandleOutgoing(), recyclerView,
                        OUTGOING_SHARES_ADAPTER, MegaNodeAdapter.ITEM_VIEW_TYPE_LIST, sortByHeaderViewModel);
            }
        } else {
            v = getGridView(inflater, container);

            if (adapter == null) {
                adapter = new MegaNodeAdapter(context, this, nodes,
                        managerActivity.getParentHandleOutgoing(), recyclerView,
                        OUTGOING_SHARES_ADAPTER, MegaNodeAdapter.ITEM_VIEW_TYPE_GRID, sortByHeaderViewModel);
            }

            gridLayoutManager.setSpanSizeLookup(adapter.getSpanSizeLookup(gridLayoutManager.getSpanCount()));
        }

        adapter.setParentHandle(managerActivity.getParentHandleOutgoing());
        adapter.setListFragment(recyclerView);

        if (managerActivity.getParentHandleOutgoing() == INVALID_HANDLE) {
            Timber.w("Parent Handle == -1");
            findNodes();
            adapter.setParentHandle(INVALID_HANDLE);
        } else {
            managerActivity.hideTabs(true, OUTGOING_TAB);
            MegaNode parentNode = megaApi.getNodeByHandle(managerActivity.getParentHandleOutgoing());
Timber.d("Parent Handle: %s",  managerActivity.getParentHandleOutgoing());

            nodes = megaApi.getChildren(parentNode, sortOrderManagement.getOrderCloud());
            adapter.setNodes(nodes);
        }

        managerActivity.setToolbarTitle();
        managerActivity.supportInvalidateOptionsMenu();

        adapter.setMultipleSelect(false);
        recyclerView.setAdapter(adapter);
        visibilityFastScroller();
        setEmptyView();

        return v;
    }

    @Override
    public void refresh() {
Timber.d("Parent Handle: %s",  managerActivity.getParentHandleOutgoing());

        if (managerActivity.getParentHandleOutgoing() == -1) {
            findNodes();
        } else {
            MegaNode n = megaApi.getNodeByHandle(managerActivity.getParentHandleOutgoing());
            managerActivity.setToolbarTitle();

            nodes = megaApi.getChildren(n, sortOrderManagement.getOrderCloud());
            adapter.setNodes(nodes);
        }

        managerActivity.supportInvalidateOptionsMenu();

        visibilityFastScroller();
        hideActionMode();
        setEmptyView();
    }

    public void findNodes() {
        ArrayList<MegaShare> outNodeList = megaApi.getOutShares();

        nodes.clear();
        long lastFolder = -1;

        for (int k = 0; k < outNodeList.size(); k++) {
            if (outNodeList.get(k).getUser() != null) {
                MegaShare mS = outNodeList.get(k);
                MegaNode node = megaApi.getNodeByHandle(mS.getNodeHandle());

                if (lastFolder != node.getHandle()) {
                    lastFolder = node.getHandle();
                    nodes.add(node);
                }
            }
        }

        orderNodes();
    }

    @Override
    public void itemClick(int position) {
        if (adapter.isMultipleSelect()) {
            Timber.d("multiselect ON");

            adapter.toggleSelection(position);

            List<MegaNode> selectedNodes = adapter.getSelectedNodes();
            if (selectedNodes.size() > 0) {
                updateActionModeTitle();
            }
        } else if (nodes.get(position).isFolder()) {
            navigateToFolder(nodes.get(position));
        } else {
            //Is file
            openFile(nodes.get(position), OUTGOING_SHARES_ADAPTER, position);
        }
    }

    @Override
    public void navigateToFolder(MegaNode node) {
        managerActivity.hideTabs(true, OUTGOING_TAB);
        managerActivity.increaseDeepBrowserTreeOutgoing();
        Timber.d("deepBrowserTree after clicking folder%s", managerActivity.deepBrowserTreeOutgoing);

        int lastFirstVisiblePosition = 0;
        if (managerActivity.isList) {
            lastFirstVisiblePosition = mLayoutManager.findFirstCompletelyVisibleItemPosition();
        } else {
            lastFirstVisiblePosition = ((NewGridRecyclerView) recyclerView).findFirstCompletelyVisibleItemPosition();
            if (lastFirstVisiblePosition == -1) {
                lastFirstVisiblePosition = ((NewGridRecyclerView) recyclerView).findFirstVisibleItemPosition();
            }
        }

        lastPositionStack.push(lastFirstVisiblePosition);

        managerActivity.setParentHandleOutgoing(node.getHandle());

        managerActivity.supportInvalidateOptionsMenu();
        managerActivity.setToolbarTitle();

        nodes = megaApi.getChildren(node, sortOrderManagement.getOrderCloud());

        adapter.setNodes(nodes);
        recyclerView.scrollToPosition(0);
        visibilityFastScroller();
        setEmptyView();
        checkScroll();
        managerActivity.showFabButton();
    }

    @Override
    public void setNodes(ArrayList<MegaNode> nodes) {
        this.nodes = nodes;
        orderNodes();
    }

    protected void orderNodes() {
        if (sortOrderManagement.getOrderOthers() == MegaApiJava.ORDER_DEFAULT_DESC) {
            sortByNameDescending(this.nodes);
        } else {
            sortByNameAscending(this.nodes);
        }

        adapter.setNodes(nodes);
    }

    @Override
    public int onBackPressed() {
        Timber.d("deepBrowserTree: %s", managerActivity.deepBrowserTreeOutgoing);

        if (adapter == null) {
            return 0;
        }

        managerActivity.decreaseDeepBrowserTreeOutgoing();
        managerActivity.supportInvalidateOptionsMenu();
        if (managerActivity.deepBrowserTreeOutgoing == 0) {
            Timber.d("deepBrowserTree==0");
            //In the beginning of the navigation
            managerActivity.setParentHandleOutgoing(INVALID_HANDLE);
            managerActivity.hideTabs(false, OUTGOING_TAB);

            managerActivity.setToolbarTitle();
            findNodes();
            visibilityFastScroller();

            int lastVisiblePosition = 0;
            if (!lastPositionStack.empty()) {
                lastVisiblePosition = lastPositionStack.pop();
            }

            if (lastVisiblePosition >= 0) {
                if (managerActivity.isList) {
                    mLayoutManager.scrollToPositionWithOffset(lastVisiblePosition, 0);
                } else {
                    gridLayoutManager.scrollToPositionWithOffset(lastVisiblePosition, 0);
                }
            }

            recyclerView.setVisibility(View.VISIBLE);
            emptyImageView.setVisibility(View.GONE);
            emptyLinearLayout.setVisibility(View.GONE);
            managerActivity.showFabButton();
            return 3;
        } else if (managerActivity.deepBrowserTreeOutgoing > 0) {
            Timber.d("Keep navigation");

            MegaNode parentNode = megaApi.getParentNode(megaApi.getNodeByHandle(managerActivity.getParentHandleOutgoing()));

            if (parentNode != null) {
                recyclerView.setVisibility(View.VISIBLE);
                emptyImageView.setVisibility(View.GONE);
                emptyLinearLayout.setVisibility(View.GONE);

                managerActivity.setParentHandleOutgoing(parentNode.getHandle());

                managerActivity.setToolbarTitle();
                managerActivity.supportInvalidateOptionsMenu();

                nodes = megaApi.getChildren(parentNode, sortOrderManagement.getOrderCloud());

                adapter.setNodes(nodes);
                visibilityFastScroller();

                int lastVisiblePosition = 0;
                if (!lastPositionStack.empty()) {
                    lastVisiblePosition = lastPositionStack.pop();
                }

                if (lastVisiblePosition >= 0) {
                    if (managerActivity.isList) {
                        mLayoutManager.scrollToPositionWithOffset(lastVisiblePosition, 0);
                    } else {
                        gridLayoutManager.scrollToPositionWithOffset(lastVisiblePosition, 0);
                    }
                }

            }
            managerActivity.showFabButton();
            return 2;
        } else {
            Timber.d("Back to Cloud");
            managerActivity.deepBrowserTreeOutgoing = 0;
            return 0;
        }
    }

    @Override
    protected void setEmptyView() {
        String textToShow = null;
        MegaNode rootNode = megaApi.getRootNode();

        if (rootNode != null && rootNode.getHandle() == managerActivity.getParentHandleOutgoing()
                || managerActivity.getParentHandleOutgoing() == -1) {
            if (isScreenInPortrait(context)) {
                emptyImageView.setImageResource(R.drawable.empty_outgoing_portrait);
            } else {
                emptyImageView.setImageResource(R.drawable.empty_outgoing_landscape);
            }
            textToShow = context.getString(R.string.context_empty_outgoing);
        }

        setFinalEmptyView(textToShow);
    }

    /**
     * Method to update an item when a nickname is added, updated or removed from a contact.
     *
     * @param contactHandle Contact ID.
     */
    public void updateContact(long contactHandle) {
        adapter.updateItem(contactHandle);
    }
}

File: app/src/main/java/mega/privacy/android/app/presentation/shares/outgoing/OutgoingSharesFragment.kt
package mega.privacy.android.app.presentation.shares.outgoing

import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.view.ActionMode
import androidx.fragment.app.activityViewModels
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import androidx.recyclerview.widget.DefaultItemAnimator
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.launch
import mega.privacy.android.app.R
import mega.privacy.android.app.arch.extensions.collectFlow
import mega.privacy.android.app.components.CustomizedGridLayoutManager
import mega.privacy.android.app.components.PositionDividerItemDecoration
import mega.privacy.android.app.fragments.homepage.EventObserver
import mega.privacy.android.app.fragments.homepage.SortByHeaderViewModel
import mega.privacy.android.app.main.adapters.MegaNodeAdapter
import mega.privacy.android.app.presentation.contact.authenticitycredendials.AuthenticityCredentialsActivity
import mega.privacy.android.app.presentation.manager.model.SharesTab
import mega.privacy.android.app.presentation.manager.model.Tab
import mega.privacy.android.app.presentation.shares.MegaNodeBaseFragment
import mega.privacy.android.app.utils.CloudStorageOptionControlUtil
import mega.privacy.android.app.utils.Constants
import mega.privacy.android.app.utils.Constants.ORDER_CLOUD
import mega.privacy.android.app.utils.Constants.ORDER_OTHERS
import mega.privacy.android.app.utils.MegaNodeUtil.areAllFileNodesAndNotTakenDown
import mega.privacy.android.app.utils.MegaNodeUtil.areAllNotTakenDown
import mega.privacy.android.app.utils.MegaNodeUtil.canMoveToRubbish
import mega.privacy.android.app.utils.Util
import mega.privacy.android.app.utils.displayMetrics
import mega.privacy.android.domain.entity.ShareData
import mega.privacy.android.domain.entity.SortOrder
import mega.privacy.android.domain.entity.preference.ViewType
import nz.mega.sdk.MegaApiJava.INVALID_HANDLE
import nz.mega.sdk.MegaError
import nz.mega.sdk.MegaNode
import nz.mega.sdk.MegaShare
import timber.log.Timber

/**
 * Outgoing shares page
 */
@AndroidEntryPoint
class OutgoingSharesFragment : MegaNodeBaseFragment() {

    private val viewModel by activityViewModels<OutgoingSharesViewModel>()

    private fun outgoingSharesState() = viewModel.state.value

    private val itemDecoration: PositionDividerItemDecoration by lazy(LazyThreadSafetyMode.NONE) {
        PositionDividerItemDecoration(requireContext(), displayMetrics())
    }

    /**
     * onCreateView
     */
    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?,
    ): View? {
        super.onCreateView(inflater, container, savedInstanceState)

        if (megaApi.rootNode == null)
            return null

        val view = setupUI(inflater, container)
        setupAdapter()
        switchViewType()

        return view
    }

    /**
     * onViewCreated
     */
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        setupObservers()
    }

    /**
     * activateActionMode
     */
    override fun activateActionMode() {
        if (megaNodeAdapter?.isMultipleSelect == true) return

        super.activateActionMode()
        actionMode =
            (requireActivity() as AppCompatActivity).startSupportActionMode(
                ActionBarCallBack(SharesTab.OUTGOING_TAB)
            )
    }

    override fun itemClick(position: Int) {
        val actualPosition = position - 1
        val node = outgoingSharesState().nodes.getOrNull(actualPosition)?.first
        val shareData = outgoingSharesState().nodes.getOrNull(actualPosition)?.second
        when {
            shareData?.isPending == true -> {
                showCanNotVerifyContact(shareData.user)
            }
            shareData?.isVerified == false -> {
                openAuthenticityCredentials(shareData.user)
            }
            // select mode
            megaNodeAdapter?.isMultipleSelect == true -> {
                megaNodeAdapter?.toggleSelection(position)
                val selectedNodes = megaNodeAdapter?.selectedNodes
                if ((selectedNodes?.size ?: 0) > 0)
                    updateActionModeTitle()
            }

            // click on a folder
            node?.isFolder == true ->
                navigateToFolder(node)

            // click on a file
            else ->
                node?.let {
                    openFile(
                        it,
                        Constants.OUTGOING_SHARES_ADAPTER,
                        actualPosition
                    )
                }
        }
    }

    override fun navigateToFolder(node: MegaNode) {
        Timber.d("Is folder deep: %s", outgoingSharesState().outgoingTreeDepth)

        val lastFirstVisiblePosition: Int = when {
            outgoingSharesState().currentViewType == ViewType.LIST ->
                recyclerView?.findFirstCompletelyVisibleItemPosition() ?: 0

            recyclerView?.findFirstCompletelyVisibleItemPosition() == -1 ->
                recyclerView?.findFirstVisibleItemPosition() ?: 0

            else ->
                recyclerView?.findFirstCompletelyVisibleItemPosition() ?: 0
        }

        viewModel.pushToLastPositionStack(lastFirstVisiblePosition)
        viewModel.increaseOutgoingTreeDepth(node.handle)
        recyclerView?.scrollToPosition(0)
        checkScroll()
    }

    override fun onBackPressed(): Int {
        Timber.d("deepBrowserTree: %s", managerActivity?.deepBrowserTreeOutgoing)

        if (megaNodeAdapter == null)
            return 0

        managerActivity?.invalidateOptionsMenu()

        return when {
            outgoingSharesState().outgoingTreeDepth == 1 -> {
                Timber.d("deepBrowserTree==1")
                viewModel.resetOutgoingTreeDepth()

                val lastVisiblePosition = viewModel.popLastPositionStack()

                lastVisiblePosition.takeIf { it > 0 }?.let {
                    recyclerView?.scrollToPosition(it)
                }

                recyclerView?.visibility = View.VISIBLE
                emptyListImageView?.visibility = View.GONE
                3
            }

            outgoingSharesState().outgoingTreeDepth > 1 -> {
                Timber.d("deepTree>1")

                outgoingSharesState().outgoingParentHandle?.let { parentHandle ->
                    recyclerView?.visibility = View.VISIBLE
                    emptyListImageView?.visibility = View.GONE
                    viewModel.decreaseOutgoingTreeDepth(parentHandle)

                    val lastVisiblePosition = viewModel.popLastPositionStack()

                    lastVisiblePosition.takeIf { it > 0 }?.let {
                        recyclerView?.scrollToPosition(it)
                    }
                }

                2
            }

            else -> {
                Timber.d("ELSE deepTree")
                viewModel.resetOutgoingTreeDepth()
                0
            }
        }
    }

    override fun showSortByPanel() {
        val orderType = when (outgoingSharesState().outgoingTreeDepth) {
            0 -> ORDER_OTHERS
            else -> ORDER_CLOUD
        }
        managerActivity?.showNewSortByPanel(orderType)
    }

    override val viewerFrom: Int = Constants.VIEWER_FROM_OUTGOING_SHARES
    override val currentSharesTab: SharesTab = SharesTab.OUTGOING_TAB
    override val sortOrder: SortOrder
        get() = outgoingSharesState().sortOrder
    override val parentHandle: Long
        get() = outgoingSharesState().outgoingHandle

    /**
     * Setup ViewModel observers
     */
    private fun setupObservers() {
        viewLifecycleOwner.lifecycleScope.launch {
            viewLifecycleOwner.repeatOnLifecycle(Lifecycle.State.RESUMED) {
                viewModel.state.collect {
                    Timber.d("Collect ui state")

                    // If the nodes are loading, don't display the UI
                    if (it.isLoading) {
                        recyclerView?.visibility = View.GONE
                        hideTabs(true)
                        return@collect
                    }

                    hideTabs(!it.isFirstNavigationLevel())

                    managerActivity?.showFabButton()
                    managerActivity?.invalidateOptionsMenu()
                    managerActivity?.setToolbarTitle()

                    visibilityFastScroller()
                    hideActionMode()
                    updateNodes(it.nodes)
                    setEmptyView(it.isInvalidHandle)
                }
            }
        }

        sortByHeaderViewModel.showDialogEvent.observe(viewLifecycleOwner, EventObserver {
            showSortByPanel()
        })

        viewLifecycleOwner.collectFlow(sortByHeaderViewModel.state) { state ->
            updateViewType(state.viewType)
        }
    }

    /**
     * Updates the View Type of this Fragment
     *
     * Changing the View Type will cause the scroll position to be lost. To avoid that, only
     * refresh the contents when the new View Type is different from the original View Type
     *
     * @param viewType The new View Type received from [SortByHeaderViewModel]
     */
    private fun updateViewType(viewType: ViewType) {
        if (viewType != outgoingSharesState().currentViewType) {
            viewModel.setCurrentViewType(viewType)
            switchViewType()
        }
    }

    /**
     * Switches how items in the [MegaNodeAdapter] are being displayed, based on the current
     * [ViewType] in [OutgoingSharesViewModel]
     */
    private fun switchViewType() {
        recyclerView?.run {
            when (outgoingSharesState().currentViewType) {
                ViewType.LIST -> {
                    switchToLinear()
                    itemAnimator = Util.noChangeRecyclerViewItemAnimator()
                    if (itemDecorationCount == 0) addItemDecoration(itemDecoration)
                    megaNodeAdapter?.adapterType = MegaNodeAdapter.ITEM_VIEW_TYPE_LIST
                }
                ViewType.GRID -> {
                    switchBackToGrid()
                    itemAnimator = DefaultItemAnimator()
                    removeItemDecoration(itemDecoration)
                    (layoutManager as CustomizedGridLayoutManager).apply {
                        spanSizeLookup = megaNodeAdapter?.getSpanSizeLookup(spanCount)
                    }
                    megaNodeAdapter?.adapterType = MegaNodeAdapter.ITEM_VIEW_TYPE_GRID
                }
            }
        }
    }

    /**
     * Update displayed nodes
     *
     * @param nodes the list of nodes to display with his shareData associated
     */
    private fun updateNodes(nodes: List<Pair<MegaNode, ShareData?>>) {
        val mutableListNodes = nodes.map { it.first }
        val mutableListShareData = nodes.map { it.second }
        megaNodeAdapter?.setNodesWithShareData(mutableListNodes, mutableListShareData)
    }

    /**
     * Initialize the adapter
     */
    private fun setupAdapter() {
        megaNodeAdapter = MegaNodeAdapter(
            requireActivity(),
            this,
            outgoingSharesState().nodes.map { it.first },
            outgoingSharesState().outgoingHandle,
            recyclerView,
            Constants.OUTGOING_SHARES_ADAPTER,
            if (outgoingSharesState().currentViewType == ViewType.LIST) MegaNodeAdapter.ITEM_VIEW_TYPE_LIST
            else MegaNodeAdapter.ITEM_VIEW_TYPE_GRID,
            sortByHeaderViewModel
        ).also {
            it.isMultipleSelect = false
            recyclerView?.adapter = it
        }
    }

    /**
     * Hide/Show shares tab
     *
     * @param hide true if needs to hide shares tabs
     */
    private fun hideTabs(hide: Boolean) {
        managerActivity?.hideTabs(hide, SharesTab.OUTGOING_TAB)
    }

    /**
     * Set the empty view and message depending if the handle is valid or not
     *
     * @param isInvalidHandle true if the handle is invalid
     */
    private fun setEmptyView(isInvalidHandle: Boolean) {
        var textToShow: String? = null

        if (isInvalidHandle) {
            if (Util.isScreenInPortrait(requireContext())) {
                emptyListImageView?.setImageResource(R.drawable.empty_outgoing_portrait)
            } else {
                emptyListImageView?.setImageResource(R.drawable.empty_outgoing_landscape)
            }
            textToShow = requireContext().getString(R.string.context_empty_outgoing)
        }
        setFinalEmptyView(textToShow)
    }

    private inner class ActionBarCallBack(currentTab: Tab) : BaseActionBarCallBack(currentTab) {
        override fun onPrepareActionMode(actionMode: ActionMode, menu: Menu): Boolean {
            super.onPrepareActionMode(actionMode, menu)
            val control = CloudStorageOptionControlUtil.Control()
            if (selected.size == 1 && !selected[0].isTakenDown) {
                if (megaApi.checkAccessErrorExtended(selected[0], MegaShare.ACCESS_OWNER).errorCode
                    == MegaError.API_OK
                ) {
                    if (selected[0].isExported) {
                        control.manageLink().setVisible(true).showAsAction =
                            MenuItem.SHOW_AS_ACTION_ALWAYS
                        control.removeLink().isVisible = true
                    } else {
                        control.link.setVisible(true).showAsAction = MenuItem.SHOW_AS_ACTION_ALWAYS
                    }
                }
                if (selected[0].isFolder) {
                    control.shareFolder().setVisible(true).showAsAction =
                        MenuItem.SHOW_AS_ACTION_ALWAYS
                }
                if (selected[0].isOutShare) {
                    control.removeShare().isVisible = true
                }
            }
            val areAllNotTakenDown = selected.areAllNotTakenDown()
            if (areAllNotTakenDown) {
                if (outgoingSharesState().outgoingHandle == INVALID_HANDLE) {
                    control.removeShare().setVisible(true).showAsAction =
                        MenuItem.SHOW_AS_ACTION_ALWAYS
                }
                control.shareOut().setVisible(true).showAsAction = MenuItem.SHOW_AS_ACTION_ALWAYS
                if (outgoingSharesState().outgoingTreeDepth > 0) {
                    if (areAllFileNodesAndNotTakenDown(selected)) {
                        control.sendToChat().setVisible(true).showAsAction =
                            MenuItem.SHOW_AS_ACTION_ALWAYS
                    }
                }
                control.copy().isVisible = true
                if (control.alwaysActionCount() < CloudStorageOptionControlUtil.MAX_ACTION_COUNT) {
                    control.copy().showAsAction = MenuItem.SHOW_AS_ACTION_ALWAYS
                } else {
                    control.copy().showAsAction = MenuItem.SHOW_AS_ACTION_NEVER
                }
            } else {
                control.saveToDevice().isVisible = false
            }
            if (selected.size == 1
                && megaApi.checkAccessErrorExtended(
                    selected[0],
                    MegaShare.ACCESS_FULL
                ).errorCode == MegaError.API_OK
            ) {
                control.rename().isVisible = true
                if (control.alwaysActionCount() < CloudStorageOptionControlUtil.MAX_ACTION_COUNT) {
                    control.rename().showAsAction = MenuItem.SHOW_AS_ACTION_ALWAYS
                } else {
                    control.rename().showAsAction = MenuItem.SHOW_AS_ACTION_NEVER
                }
            }
            control.selectAll().isVisible = notAllNodesSelected()
            control.trash().isVisible = canMoveToRubbish(selected)

            // If there is at least one Backup folder found, then apply read-only restrictions
            // for all selected items
            if (selected.any { node -> megaApi.isInInbox(node) }) {
                control.rename().isVisible = false
                control.move().isVisible = false
                control.trash().isVisible = false
            }

            CloudStorageOptionControlUtil.applyControl(menu, control)
            return true
        }
    }

    /**
     * Show cannot verify contact dialog
     * @param email : Email of the user
     */
    private fun showCanNotVerifyContact(email: String?) {
        MaterialAlertDialogBuilder(
            requireContext(),
            R.style.ThemeOverlay_Mega_MaterialAlertDialog
        ).setTitle(getString(R.string.shared_items_contact_not_in_contact_list_dialog_title))
            .setMessage(
                getString(
                    R.string.shared_items_contact_not_in_contact_list_dialog_content,
                    email
                )
            )
            .setPositiveButton(
                getString(R.string.general_ok)
            ) { dialogInterface, _ -> dialogInterface.dismiss() }
            .show()
    }

    /**
     * Open authenticityCredentials screen to verify user
     * @param email : Email of the user
     */
    private fun openAuthenticityCredentials(email: String?) {
        Intent(
            requireActivity(),
            AuthenticityCredentialsActivity::class.java
        ).apply {
            putExtra(Constants.EMAIL, email)
            requireActivity().startActivity(this)
        }
    }
}
